homeassistant/components/update/__init__.py

```
@@ -0,0 +1,297 @@
+                "icon_url": description.icon_url,
+                "provides_backup": description.provides_backup,
+            }
+            for domain, domain_data in self._updates.items()
+            if domain_data is not None
+            for description in domain_data
+            if f"{domain}_{description.identifier}_{description.available_version}"
+            not in self._skip
+        ]
+
+    async def load(self) -> None:
+        """Load the stored data."""
+        self._skip = set(await self._store.async_load() or [])
+
+    async def gather_updates(self) -> list[dict]:
+        """Gather updates."""
+        for domain, domain_data in zip(
+            self._registrations,
+            await asyncio.gather(
+                *(
+                    self._get_integration_info(registration)
+                    for registration in self._registrations.values()
+                )
+            ),
+        ):
+            self._updates[domain] = domain_data
+        return self.filtered_updates
+
+    @callback
+    def _data_to_save(self) -> list:
+        """Schedule storing the data."""
+        return list(self._skip)
+
+    def _get_update_description(
+        self, domain: str, identifier: str
+    ) -> UpdateDescription | None:
+        """Get an update."""
+        return next(
+            (
+                update
+                for update in self._updates.get(domain, [])
+                if update.identifier == identifier
+            ),
+            None,
+        )
+
+    def register(self, registration: UpdateRegistration) -> None:
+        """Register an update handler."""
+        self._registrations[registration.domain] = registration
+
+    async def perform_update(
+        self,
+        domain: str,
+        identifier: str,
+        **kwargs: Any,
+    ) -> bool | None:
+        """Perform an update."""
+        update_description = self._get_update_description(domain, identifier)
+
+        if update_description is None:
```

Let's raise for passing in invalid update.